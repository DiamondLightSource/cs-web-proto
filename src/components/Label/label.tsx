import React from "react";

import classes from "./label.module.css";
import { Component, Widget, WidgetPropType } from "../Widget/widget";
import { registerWidget } from "../register";
import { BoolPropOpt, StringOrNumProp, InferWidgetProps } from "../propTypes";

const LabelProps = {
  text: StringOrNumProp,
  visible: BoolPropOpt,
  transparent: BoolPropOpt
};

export const LabelComponent = (
  props: InferWidgetProps<typeof LabelProps> & Component
): JSX.Element => {
  const style: any = { ...props.style };
  if (props.visible !== undefined && !props.visible) {
    style["visibility"] = "hidden";
  }
  if (props.transparent !== undefined && props.transparent) {
    style["backgroundColor"] = "transparent";
  }
  // Simple component to display text - defaults to black text and dark grey background
  return (
    <div className={`Label ${classes.Label}`} style={style}>
      {props.text}
    </div>
  );
};

const LabelWidgetProps = {
  ...LabelProps,
  ...WidgetPropType
};

export const Label = (
  props: InferWidgetProps<typeof LabelWidgetProps>
): JSX.Element => <Widget baseWidget={LabelComponent} {...props} />;

registerWidget(Label, LabelWidgetProps, "label");
