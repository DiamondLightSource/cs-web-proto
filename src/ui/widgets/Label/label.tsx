import React from "react";

import classes from "./label.module.css";
import { Widget } from "../widget";
import { WidgetPropType } from "../widgetProps";
import { registerWidget } from "../register";
import {
  BoolPropOpt,
  StringProp,
  InferWidgetProps,
  StringPropOpt,
  ChoicePropOpt,
  FontPropOpt,
  ColorPropOpt,
  BorderPropOpt
} from "../propTypes";

const LabelProps = {
  text: StringProp,
  visible: BoolPropOpt,
  transparent: BoolPropOpt,
  className: StringPropOpt,
  textAlign: ChoicePropOpt(["left", "center", "right"]),
  font: FontPropOpt,
  foregroundColor: ColorPropOpt,
  backgroundColor: ColorPropOpt,
  border: BorderPropOpt
};

const LabelWidgetProps = {
  ...LabelProps,
  ...WidgetPropType
};

type LabelWidget = InferWidgetProps<typeof LabelWidgetProps>;

export const LabelComponent = (
  props: InferWidgetProps<typeof LabelProps>
): JSX.Element => {
  const { visible = true, transparent = false, textAlign = "center" } = props;
  const className = props.className ?? `Label ${classes.Label}`;
  const style: any = { ...props.font?.asStyle(), ...props.border?.asStyle() };
  if (!visible) {
    style["visibility"] = "hidden";
  }
  // Since display is "flex", use "flex-start" and "flex-end" to align
  // the content.
  let alignment = "center";
  if (textAlign === "left") {
    alignment = "flex-start";
  } else if (textAlign === "right") {
    alignment = "flex-end";
  }
  style["justifyContent"] = alignment;
  style["color"] = props.foregroundColor?.rgbaString();
  style["backgroundColor"] = props.backgroundColor?.rgbaString();
  // Transparent prop overrides backgroundColor.
  if (transparent) {
    style["backgroundColor"] = "transparent";
  }
  // Simple component to display text - defaults to black text and dark grey background
  return (
    <div className={className} style={style}>
      {props.text}
    </div>
  );
};

export const Label = (props: LabelWidget): JSX.Element => (
  <Widget baseWidget={LabelComponent} {...props} />
);

registerWidget(Label, LabelWidgetProps, "label");
