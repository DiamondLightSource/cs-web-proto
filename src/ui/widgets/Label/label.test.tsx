import React from "react";
import { LabelComponent } from "./label";
import { shallow, ShallowWrapper } from "enzyme";
import { create, ReactTestRenderer } from "react-test-renderer";

let snapshot: ReactTestRenderer;
let wrapper: ShallowWrapper;

beforeEach((): void => {
  const label = <LabelComponent text="hello" />;
  wrapper = shallow(label);
  snapshot = create(label);
});

describe("<Label Label />", (): void => {
  test("it matches the snapshot", (): void => {
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test("it renders a basic element", (): void => {
    expect(wrapper.text()).toEqual("hello");
  });

  test("it handles transparent prop", (): void => {
    wrapper = shallow(<LabelComponent text={"hello"} transparent={true} />);
    expect(
      wrapper.find("div").getElement().props.style["backgroundColor"]
    ).toEqual("transparent");
  });
});
