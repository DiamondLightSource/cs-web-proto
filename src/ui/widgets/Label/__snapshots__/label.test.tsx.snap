// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Label /> it matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="Label Label"
    style="cursor: auto; justify-content: center;"
  >
    hello
  </div>
</DocumentFragment>
`;
